src/Apple/Provider.php

```diff
@@ -0,0 +1,204 @@
+    public function getAccessTokenResponse($code)
+    {
+        $response = $this->getHttpClient()->post($this->getTokenUrl(), [
+            'headers'        => ['Authorization' => 'Basic '.base64_encode($this->clientId.':'.$this->clientSecret)],
+            'form_params'    => $this->getTokenFields($code),
+        ]);
+
+        return json_decode($response->getBody(), true);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getTokenFields($code)
+    {
+        return array_merge(parent::getTokenFields($code), [
+            'grant_type' => 'authorization_code',
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getUserByToken($token)
+    {
+        static::verify($token);
+        $claims = explode('.', $token)[1];
+
+        return json_decode(base64_decode($claims), true);
+    }
+
+    /**
+     * Verify apple jwt.
+     *
+     * @param $jwt
+     *
+     * @return object
+     *
+     * @see https://appleid.apple.com/auth/keys
+     */
+    public static function verify($jwt)
+    {
+        $signer = new Sha256();
+
+        $token = (new Parser())->parse((string) $jwt);
+
+        if ($token->getClaim('iss') !== self::URL) {
+            throw new InvalidStateException('Invalid Issuer', Response::HTTP_UNAUTHORIZED);
+        }
+        if ($token->isExpired()) {
+            throw new InvalidStateException('Token Expired', Response::HTTP_UNAUTHORIZED);
+        }
+
+        $data = json_decode(file_get_contents(self::URL.'/auth/keys'), true);
+
+        $public_keys = JWK::parseKeySet($data);
+
+        $signature_verified = false;
+
+        foreach ($public_keys as $res) {
+            $publicKey = openssl_pkey_get_details($res);
+            if ($token->verify($signer, $publicKey['key'])) {
+                $signature_verified = true;
+            }
+        }
+        if (!$signature_verified) {
+            throw new InvalidStateException('Invalid JWT Signature');
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function user()
+    {
+        //Temporary fix to enable stateless
+        if ($this->usesState()) {
+            $this->request->session()->put('state', $this->request->input('state'));
+            if ($this->hasInvalidState()) {
+                throw new InvalidStateException();
+            }
+        }
+
+        $response = $this->getAccessTokenResponse($this->getCode());
+
+        $user = $this->mapUserToObject($this->getUserByToken(
+            $token = Arr::get($response, 'id_token')
+        ));
+
+        if ($user instanceof User) {
+            $user->setAccessTokenResponseBody($response);
+        }
+
+        return $user->setToken($token)
+            ->setRefreshToken(Arr::get($response, 'refresh_token'))
+            ->setExpiresIn(Arr::get($response, 'expires_in'));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function mapUserToObject(array $user)
+    {
+        if (request()->filled('user')) {
```

Comment:
Why are we reading from the request here? Shouldn't this come from the JWT?

Reply:
According to the [documentation](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_js/incorporating_sign_in_with_apple_into_other_platforms), the available valid values for scope(name and email) are sent as a user JSON object in form_post.

Reply:
According to the documentation, the available valid values for scope(name and email) are sent as a user JSON object in form_post.